# OpenVLA 复现课程笔记

这份 notebook 按照「硬件检测 → 固定 12/4 episode 划分 → 12-episode QLoRA 训练 → adapter 验证 → held-out ground-truth 评测」的顺序，记录 OpenVLA-7B + QLoRA 的复现过程。

全流程只使用 `train[:12]` 训练、`train[12:16]` 评测，训练数据和 held-out 数据完全不重叠。


In [ ]:
import os

# ==== 全局实验配置 ====
# 本 notebook 从第一步起就固定为「12 个训练 episode + 4 个 held-out episode」，
# 后续训练、adapter smoke test、ground-truth 评测都复用这里的取值。
os.environ['OPENVLA_TRAIN_EPISODES'] = '12'
os.environ['OPENVLA_EVAL_EPISODE_START'] = '12'
os.environ['OPENVLA_EVAL_EPISODES'] = '4'
os.environ['OPENVLA_RUN_ROOT'] = 'runs/libero_spatial_12episode_1000step'
os.environ['OPENVLA_ADAPTER_ROOT'] = 'adapter-tmp/libero_spatial_12episode_1000step'

# ==== 离线运行 ====
os.environ['HF_HUB_OFFLINE'] = '1'
os.environ['TRANSFORMERS_OFFLINE'] = '1'
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

TRAIN_EPISODES = int(os.environ['OPENVLA_TRAIN_EPISODES'])
EVAL_EPISODE_START = int(os.environ['OPENVLA_EVAL_EPISODE_START'])
EVAL_EPISODES = int(os.environ['OPENVLA_EVAL_EPISODES'])
RUN_ROOT = os.environ['OPENVLA_RUN_ROOT']
ADAPTER_ROOT = os.environ['OPENVLA_ADAPTER_ROOT']

print('训练划分    :', f'train[:{TRAIN_EPISODES}]')
print('held-out    :', f'train[{EVAL_EPISODE_START}:{EVAL_EPISODE_START + EVAL_EPISODES}]')
print('run 目录    :', RUN_ROOT)
print('adapter 目录:', ADAPTER_ROOT)


## 1. 硬件设备检测

先确认操作系统、Python、GPU、显存和 CUDA 是否正常。

In [8]:
import os
import platform
import subprocess
import torch

print('操作系统:', platform.system(), platform.release())
print('Python 版本:', platform.python_version())
print('Conda 环境:', os.environ.get('CONDA_DEFAULT_ENV', 'not-set'))
print('CUDA 可用:', torch.cuda.is_available())
print('CUDA 版本:', torch.version.cuda)
print('GPU 数量:', torch.cuda.device_count())
if torch.cuda.is_available():
    print('当前 GPU:', torch.cuda.get_device_name(0))
    print('GPU 显存:', round(torch.cuda.get_device_properties(0).total_memory / 1024**3, 2), 'GB')
try:
    print(subprocess.check_output(['nvidia-smi', '--query-gpu=name,driver_version,memory.total,memory.free', '--format=csv,noheader'], text=True))
except Exception as error:
    print('nvidia-smi 不可用:', error)

操作系统: Linux 6.8.0-138-generic
Python 版本: 3.10.20
Conda 环境: openvla
CUDA 可用: True
CUDA 版本: 12.8
GPU 数量: 1
当前 GPU: NVIDIA GeForce RTX 5070 Ti Laptop GPU
GPU 显存: 11.47 GB
NVIDIA GeForce RTX 5070 Ti Laptop GPU, 595.91.07, 12227 MiB, 2638 MiB



## 2. Python 环境与依赖检查

确认 OpenVLA 所需的主要依赖可以正常导入。

In [9]:
import importlib

required_packages = ['torch', 'transformers', 'peft', 'bitsandbytes', 'PIL', 'numpy', 'pandas', 'ipykernel']
for package_name in required_packages:
    try:
        module = importlib.import_module(package_name)
        print(f'{package_name}: OK -> {getattr(module, "__version__", "unknown")}')
    except Exception as error:
        print(f'{package_name}: MISSING -> {error}')

torch: OK -> 2.8.0+cu128
transformers: OK -> 4.40.1
peft: OK -> 0.11.1
bitsandbytes: OK -> 0.50.2
PIL: OK -> 12.3.0
numpy: OK -> 2.2.5
pandas: OK -> 2.3.3
ipykernel: OK -> 7.3.0


## 3. 仓库状态与关键文件

确认复现入口、训练脚本、adapter 和文档目录都存在。

In [10]:
import os
from pathlib import Path

repo_candidates = []
if os.environ.get('OPENVLA_REPO'):
    repo_candidates.append(Path(os.environ['OPENVLA_REPO']).expanduser())

current_path = Path.cwd().expanduser().resolve()
repo_candidates.extend([current_path, *current_path.parents])

repo = next(
    (
        candidate.resolve()
        for candidate in repo_candidates
        if (candidate / 'reproduction' / 'test_planB_libero_compare.py').exists()
        and (candidate / 'vla-scripts' / 'finetune.py').exists()
    ),
    None,
)
if repo is None:
    raise FileNotFoundError(
        '无法定位 OpenVLA 仓库。请设置 OPENVLA_REPO，或从仓库目录打开 notebook。'
    )

key_paths = [
    'reproduction/run_qlora_1000.py',
    'vla-scripts/finetune.py',
    'reproduction/test_1000step_adapter.py',
    'reproduction/test_planB_libero_compare.py',
    'reproduction/test_planB2_libero_gt.py',
    'adapter-tmp',
    'AIGC_document',
]
for relative_path in key_paths:
    path = repo / relative_path
    print(relative_path, '-> exists' if path.exists() else '-> missing')
print('仓库根目录:', repo)

reproduction/run_qlora_1000.py -> exists
vla-scripts/finetune.py -> exists
reproduction/test_1000step_adapter.py -> exists
reproduction/test_planB_libero_compare.py -> exists
reproduction/test_planB2_libero_gt.py -> exists
adapter-tmp -> exists
AIGC_document -> exists
仓库根目录: /home/yyz/openvla


## 4. 单机兼容性修复

主要修改集中在训练入口，而不是 OpenVLA 基础模型结构。单卡运行时只在分布式进程组已初始化时使用 DDP 和 barrier；保存模型时同时兼容 DDP 包装对象和普通模型对象。

In [11]:
def unwrap_model_for_saving(model):
    return model.module if hasattr(model, 'module') else model

print('单机模式：不包装 DDP，不调用 dist.barrier()')
print('分布式模式：使用 DDP，并通过 model.module 保存原始模型')

单机模式：不包装 DDP，不调用 dist.barrier()
分布式模式：使用 DDP，并通过 model.module 保存原始模型


## 5. 本地缓存与离线运行

使用本地 Hugging Face 缓存，避免 Hub 网络访问影响复现。
本地复现依赖本地模型缓存、离线环境和可用显存。验证 adapter 效果时不需要强制 merge。


In [ ]:
import os
from pathlib import Path

os.environ['HF_HUB_OFFLINE'] = '1'
os.environ['TRANSFORMERS_OFFLINE'] = '1'
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

adapter_root = os.environ.get('OPENVLA_ADAPTER_ROOT', 'adapter-tmp/libero_spatial_12episode_1000step')

hf_home = Path(os.environ.get('HF_HOME', Path.home() / '.cache' / 'huggingface'))
base_cache = hf_home / 'hub' / 'models--openvla--openvla-7b' / 'snapshots'
adapter_dir = repo / adapter_root
print('仓库路径:', repo)
print('base 模型缓存存在:', base_cache.exists())
print('adapter 根目录:', adapter_dir)
print('adapter 目录存在:', adapter_dir.exists())
print('HF_HUB_OFFLINE =', os.environ['HF_HUB_OFFLINE'])
print('TRANSFORMERS_OFFLINE =', os.environ['TRANSFORMERS_OFFLINE'])


## 6. 显存检查

加载 4-bit 模型前，先确认没有其他 Python、Jupyter 或实验进程占用大量显存。

In [13]:
import subprocess

try:
    output = subprocess.check_output(
        ['nvidia-smi', '--query-compute-apps=pid,process_name,used_memory', '--format=csv,noheader'],
        text=True,
    )
    print(output or '当前没有计算进程')
except Exception as error:
    print('无法查看 GPU 进程:', error)

12488, /home/yyz/miniconda3/envs/fast3r/bin/python, 8020 MiB



## 7. 固定 12/4 episode 划分（实验配置）

数据划分不是 notebook 末尾补做的步骤，而是整个实验的前置配置：

- 训练：`train[:12]`（episode 0-11）
- held-out：`train[12:16]`（episode 12-15）

本地 `libero_spatial_no_noops` 共 432 个 episode、16 个 TFRecord 分片（每片 27 个 episode），所以「16 个分片」并不等于「16 个 episode」。这里按 TFDS 的真实 episode 索引划分，保证训练数据与评测数据完全不重叠。

下一节训练通过 `OPENVLA_TRAIN_EPISODES=12` 让 `prismatic/vla/datasets/rlds/dataset.py` 只读取 `train[:12]`；评测脚本通过 `OPENVLA_EVAL_EPISODE_START=12` 与 `OPENVLA_EVAL_EPISODES=4` 只读取 `train[12:16]`。


In [ ]:
import json
import os
from pathlib import Path

train_episodes = int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12'))
eval_start = int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12'))
eval_episodes = int(os.environ.get('OPENVLA_EVAL_EPISODES', '4'))

repo_candidates = []
if os.environ.get('OPENVLA_REPO'):
    repo_candidates.append(Path(os.environ['OPENVLA_REPO']).expanduser())
current_path = Path.cwd().expanduser().resolve()
repo_candidates.extend([current_path, *current_path.parents])
repo = next(
    (
        candidate.resolve()
        for candidate in repo_candidates
        if (candidate / 'reproduction' / 'test_planB_libero_compare.py').exists()
        and (candidate / 'vla-scripts' / 'finetune.py').exists()
    ),
    None,
)
if repo is None:
    raise FileNotFoundError('无法定位 OpenVLA 仓库，请设置 OPENVLA_REPO。')

manifest = {
    'dataset': 'libero_spatial_no_noops',
    'dataset_total_episodes': 432,
    'dataset_tfrecord_shards': 16,
    'train_split': f'train[:{train_episodes}]',
    'train_episode_range': [0, train_episodes],
    'held_out_split': f'train[{eval_start}:{eval_start + eval_episodes}]',
    'held_out_episode_range': [eval_start, eval_start + eval_episodes],
    'seed': 42,
    'strict_use': 'Retrain on train[:12]; evaluate only on train[12:16].',
}

manifest_path = repo / 'reproduction' / 'held_out_split.json'
manifest_path.write_text(json.dumps(manifest, ensure_ascii=False, indent=2), encoding='utf-8')
print(json.dumps(manifest, ensure_ascii=False, indent=2))
print('清单已保存:', manifest_path)


## 8. 数据划分校验（TFDS episode 切片）

训练前确认两件事：数据集的真实规模，以及 `train[:12]` 与 `train[12:16]` 是否真的落在不同的 episode 上。校验通过后再启动训练，避免把评测数据混进训练集。


In [ ]:
import os
from pathlib import Path

import tensorflow_datasets as tfds

train_episodes = int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12'))
eval_start = int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12'))
eval_episodes = int(os.environ.get('OPENVLA_EVAL_EPISODES', '4'))

dataset_root = Path(os.environ.get('LIBERO_RLDS_ROOT', Path.home() / 'modified_libero_rlds'))
dataset_name = 'libero_spatial_no_noops'

builder = tfds.builder(dataset_name, data_dir=str(dataset_root))
split_info = builder.info.splits['train']
print('数据集        :', dataset_name)
print('总 episode    :', split_info.num_examples)
print('TFRecord 分片 :', split_info.num_shards)
print('每片 episode  :', list(split_info.shard_lengths))
print('是否存在 val  :', 'val' in builder.info.splits)

train_split = f'train[:{train_episodes}]'
held_out_split = f'train[{eval_start}:{eval_start + eval_episodes}]'

for split in (train_split, held_out_split):
    dataset = tfds.load(dataset_name, data_dir=str(dataset_root), split=split, shuffle_files=False)
    episode_count = 0
    first_instruction = None
    for example in dataset:
        if first_instruction is None:
            first_step = next(iter(example['steps'].as_numpy_iterator()))
            first_instruction = first_step['language_instruction'].decode('utf-8', errors='replace')
        episode_count += 1
    print(f'{split}: {episode_count} 个 episode | 首条指令: {first_instruction}')


## 9. LoRA 微调（启动脚本已参数化）

`reproduction/run_qlora_1000.py` 现在支持用参数选择数据规模、步数和输出目录，默认值仍是 12-episode 实验：

```bash
# 12-episode 实验（第 1 节里的默认配置）
python reproduction/run_qlora_1000.py

# 论文口径：官方全量 LIBERO-Spatial（432 条 demo）+ 约 1 个 epoch
python reproduction/run_qlora_1000.py \
  --train-episodes 432 --max-steps 3500 --save-steps 500 \
  --run-name libero_spatial_full_paper
```

超参与官方一致：4-bit NF4 量化 + LoRA rank 32 / all-linear + lr 5e-4 + batch size 2 × 梯度累积 8（等效 batch 16）+ image_aug。
脚本会自动用 `systemd-inhibit --what=idle:sleep:handle-lid-switch` 阻止空闲休眠与合盖休眠——笔记本一旦 suspend，CUDA 上下文会失效、训练进程会卡死（已实际发生过一次）。

输出路径规则：`runs/<run-name>/<exp_id>/` 与 `adapter-tmp/<run-name>/<exp_id>/`，日志为 `reproduction/finetune_<run-name>.log`。

各次训练的实际结果见第 15 / 18 节。


In [ ]:
import os
import subprocess
import sys

RUN_TRAINING = False  # 改成 True 才会启动 12-episode / 1000-step 长任务

environment = os.environ.copy()
environment['OPENVLA_TRAIN_EPISODES'] = str(int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12')))
environment['OPENVLA_EVAL_EPISODE_START'] = str(int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12')))
environment['OPENVLA_EVAL_EPISODES'] = str(int(os.environ.get('OPENVLA_EVAL_EPISODES', '4')))

if RUN_TRAINING:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'run_qlora_1000.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-4000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过长时间训练；需要时将 RUN_TRAINING 改为 True。')


## 10. adapter smoke test

执行 12-episode / 1000-step adapter 的最小推理验证，确认 4-bit base 模型、LoRA adapter 和推理流程都能工作。

（下面的输出是上一版全量数据 adapter 的记录，重新训练后执行本节会覆盖为新的 adapter 结果。）


In [14]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_ADAPTER_SMOKE = False  # 改成 True 才会加载 7B + LoRA 并运行推理
if RUN_ADAPTER_SMOKE:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_1000step_adapter.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout)
    if result.stderr:
        print(result.stderr)
else:
    print('安全模式：跳过 adapter smoke test；需要时将 RUN_ADAPTER_SMOKE 改为 True。')

returncode: 1
OpenVLA 1000-step QLoRA adapter smoke test

[1/5] Loading processor...
Processor loaded.

[2/5] Loading 4-bit base OpenVLA...

I0000 00:00:1789882248.821267   21472 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:00:1789882249.137304   21472 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1789882250.147328   21472 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environme

## 11. LIBERO 对比验证

运行 base model 与 adapter 的对比脚本，确认 adapter 不仅能加载，而且会改变动作输出。

（下方输出是上一版全量数据 adapter 的记录；重训后执行本节会覆盖。）


In [15]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_COMPARE = False  # 改成 True 才会运行 base/LoRA 动作对比
if RUN_COMPARE:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_planB_libero_compare.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print('CSV exists:', (repo / 'planB_libero_compare.csv').exists())
    print(result.stdout[-4000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过 base/LoRA 动作对比；需要时将 RUN_COMPARE 改为 True。')

returncode: 1
CSV exists: True
OpenVLA Plan B
Base OpenVLA vs 1000-step LIBERO Spatial LoRA
Model      : openvla/openvla-7b
Image      : /home/yyz/openvla/test_image.jpg
Adapter    : /home/yyz/openvla/adapter-tmp/libero_spatial_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug
Statistics : /home/yyz/openvla/runs/libero_spatial_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug/dataset_statistics.json
Output     : /home/yyz/openvla/planB_libero_compare.csv
[1/7] Loading LIBERO statistics...
Dataset key: libero_spatial_no_noops
q01 : [-0.7454732  -0.66160715 -0.9375     -0.10714286 -0.20678571 -0.18428572
  0.        ]
q99 : [0.9375     0.8758929  0.93214285 0.10392857 0.17678571 0.14571428
 1.        ]
mask: [ True  True  True  True  True  True False]

[2/7] Loading image...
Image size: (686, 732)

[3/7] Loading processor...
Processor loaded.

[4/7] Using OpenVLA action decoding...
Action decodi

## 12. LIBERO ground-truth 对比（held-out episode 12-15）

在 base model 与 adapter 的动作差异对比之后，继续把 adapter 输出与 LIBERO ground truth 对齐比较，用于回答：adapter 的动作是否更接近真实动作，而不仅仅是输出发生了变化。

本节的评测数据固定为 held-out episode `train[12:16]`（4 个 episode、50 个 transition），与训练用的 `train[:12]` 完全不重叠。

注意：`PeftModel.from_pretrained()` 会包装并修改传入的 base model，因此脚本在计算 Base 输出时用 `disable_adapter()`，避免 Base 和 LoRA 错误地使用同一套参数。

（下方输出同样是旧 adapter 的记录，重训后用本节重新生成 held-out 结果。）


In [16]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_GROUND_TRUTH = False  # 改成 True 才会重新生成 12/4 held-out CSV
if RUN_GROUND_TRUTH:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_planB2_libero_gt.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-5000:])
    if result.stderr:
        print(result.stderr[-3000:])
else:
    print('安全模式：跳过 held-out ground-truth 评测；需要时将 RUN_GROUND_TRUTH 改为 True。')

returncode: 1
PLAN B-2: REAL LIBERO GROUND-TRUTH ACTION COMPARISON
PyTorch: 2.8.0+cu128
CUDA: 12.8
GPU: NVIDIA GeForce RTX 5070 Ti Laptop GPU
VRAM: 11.47 GB

LIBERO action statistics:
q01 : [-0.7454732  -0.66160715 -0.9375     -0.10714286 -0.20678571 -0.18428572
  0.        ]
q99 : [0.9375     0.8758929  0.93214285 0.10392857 0.17678571 0.14571428
 1.        ]
mask: [ True  True  True  True  True  True False]

TFRecord files: 16

Reading episode metadata...
  [01/16] libero_spatial-train.tfrecord-00000-of-00016
  [02/16] libero_spatial-train.tfrecord-00001-of-00016
  [03/16] libero_spatial-train.tfrecord-00002-of-00016
  [04/16] libero_spatial-train.tfrecord-00003-of-00016
  [05/16] libero_spatial-train.tfrecord-00004-of-00016
  [06/16] libero_spatial-train.tfrecord-00005-of-00016
  [07/16] libero_spatial-train.tfrecord-00006-of-00016
  [08/16] libero_spatial-train.tfrecord-00007-of-00016
  [09/16] libero_spatial-train.tfrecord-00008-of-00016
  [10/16] libero_spatial-train.tfrecord-000

## 13. 读取评测结果

读取生成的 CSV，分别比较 base model 和 LoRA adapter 的 MAE、MSE、L2 误差。数值越小，表示动作越接近 ground truth。


In [21]:
import pandas as pd

result_path = repo / 'planB2_libero_gt.csv'
if not result_path.exists():
    print('结果文件不存在，请先执行上一节的 LIBERO ground-truth 评测。')
else:
    results = pd.read_csv(result_path)
    print('样本数:', len(results))
    metric_columns = [
        column for column in results.columns
        if column in {'base_mae', 'lora_mae', 'base_mse', 'lora_mse', 'base_l2', 'lora_l2'}
    ]
    print(results[metric_columns].mean().sort_values())
    if {'base_l2', 'lora_l2'}.issubset(results.columns):
        improved = (results['lora_l2'] < results['base_l2']).mean()
        print(f'LoRA L2 优于 base 的样本比例: {improved:.2%}')

样本数: 50
lora_mae    0.259253
lora_mse    0.270475
base_mse    0.303784
base_mae    0.308410
lora_l2     1.280291
base_l2     1.325825
dtype: float64
LoRA L2 优于 base 的样本比例: 70.00%


## 14. 多 checkpoint 对比

固定评测流程，只改变 checkpoint：

- base model
- 12-episode / 1000-step adapter
- 后续更长训练得到的 adapter

每个 checkpoint 都使用同一批 held-out 样本、指令、图像和评价指标，这样才能判断训练步数是否带来稳定改善。


In [20]:
checkpoint_root = repo / 'adapter-tmp'
checkpoints = sorted(checkpoint_root.glob('**/adapter_config.json'))
if checkpoints:
    print('发现的 adapter checkpoint:')
    for checkpoint in checkpoints:
        print(checkpoint.parent)
else:
    print('暂未发现 adapter checkpoint。完成更多训练后，再将 checkpoint 加入统一评测。')

发现的 adapter checkpoint:
/home/yyz/openvla/adapter-tmp/47a0ec7fc4ec123775a391911046cf33cf9ed83f+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug
/home/yyz/openvla/adapter-tmp/libero_spatial_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug


## 15. 实验结果（12 训练 / 4 held-out 协议）

**训练**（日志：`reproduction/finetune_12episode_1000step.log`）

- 数据：`train[:12]`，即 432 个 episode 里的前 12 个（16 个 TFRecord 分片、每片 27 个 episode）
- 配置：4-bit NF4 + LoRA rank 32 / all-linear + lr 5e-4 + batch size 2 + 梯度累积 8 + 1000 steps
- 耗时：2026-09-21 07:36 → 10:30，1000/1000 步跑完（约 2 小时 54 分，末段 10.3 s/it）
- adapter：`adapter-tmp/libero_spatial_12episode_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug/`（968 MB）
- 训练结束后不再合并成完整权重（`merge_lora_during_training=False`）：合并需要额外约 15GB 内存和 15GB 磁盘，而所有评测脚本都直接加载 LoRA adapter
- 训练期间用 `systemd-inhibit --what=idle:sleep` 阻止空闲休眠：笔记本一旦 suspend，CUDA 上下文会失效，训练进程会卡死不再前进

**Held-out 评测**（`train[12:16]`，4 个 episode、50 个 transition，训练时完全没见过）

| 指标 | Base | LoRA（12-ep / 1000-step） | 相对变化 |
|---|---|---|---|
| MAE | 0.394726 | 0.318789 | **−19.24%** |
| MSE | 0.452780 | 0.396861 | **−12.35%** |
| L2 | 1.686117 | 1.590365 | −5.68% |

逐样本比较：LoRA 更好的样本 39/50（MAE 口径 80%、MSE 78%、L2 78%）。样本覆盖 held-out 的 4 个 episode（12/13/14/15 号），与训练用的 0-11 号没有重叠，因此这是真正的 held-out 结果。


In [ ]:
import pandas as pd

result_path = repo / 'planB2_libero_gt.csv'
results = pd.read_csv(result_path)

summary = []
for metric in ('mae', 'mse', 'l2'):
    base = results[f'base_{metric}'].mean()
    lora = results[f'lora_{metric}'].mean()
    summary.append(
        {
            'metric': metric.upper(),
            'base': round(base, 6),
            'lora_12ep_1000step': round(lora, 6),
            'change_%': round(100 * (1 - lora / base), 2),
            'lora_better_samples': int((results[f'lora_{metric}'] < results[f'base_{metric}']).sum()),
            'total_samples': len(results),
        }
    )

print(pd.DataFrame(summary).to_string(index=False))


## 16. LIBERO simulator 成功率评测

动作误差只是间接指标，最终指标是 LIBERO simulator 里的任务成功率。本机第一次使用 LIBERO，需要先做一次环境准备：

```bash
# 1) 克隆并安装 LIBERO（setup.py 的 install_requires 为空，不会动 openvla 环境里的 torch）
git clone https://github.com/Lifelong-Robot-Learning/LIBERO.git ~/LIBERO
cd ~/LIBERO && pip install -e .

# 2) 对齐 mujoco 版本：mujoco>=3.2 的 pybind11 枚举会让 robosuite 1.4.1 里的
#    `joint_type in (mjJNT_HINGE, mjJNT_SLIDE)` 断言失败，固定到 3.1.6
pip install "mujoco==3.1.6"

# 3) 让 import libero 生效（editable 安装只映射了 libero.libero 子包）
echo "$HOME/LIBERO" > "$(python -c 'import site; print(site.getsitepackages()[0])')/libero_path.pth"

# 4) 非交互生成 LIBERO 路径配置（默认会在 import 时弹交互式提问）
#    ~/.libero/config.yaml 内容见仓库说明
```

另外做了三处兼容性修改（原因都写在代码注释里）：

- `~/LIBERO/libero/libero/benchmark/__init__.py`：`torch.load(..., weights_only=False)`，否则 torch>=2.6 无法加载 LIBERO 的 init states
- `experiments/robot/openvla_utils.py`：支持 LoRA adapter 目录（base 模型 + PeftModel）、processor 回退到 base 模型、用 `sdpa` 替代本机没有的 flash-attn、`norm_stats` 同时写到内部 base 模型
- `experiments/robot/libero/run_libero_eval.py`：新增 `--base_model_path`，用于指定 LoRA 对应的 base 模型

评测命令（冒烟用 `num_trials_per_task 2`，正式对齐论文用 50，即 10 个任务 × 50 次 = 500 个 rollout）：

```bash
MUJOCO_GL=egl python experiments/robot/libero/run_libero_eval.py \
  --model_family openvla \
  --pretrained_checkpoint <ADAPTER_DIR> \
  --task_suite_name libero_spatial \
  --num_trials_per_task 50 \
  --load_in_4bit True
```

结果写入 `experiments/logs/EVAL-libero_spatial-openvla-<时间>.txt`（每个任务的成功次数 + 总成功率）。


In [ ]:
import os
import subprocess
import sys

# 论文口径：10 任务 × 50 次 = 500 个 rollout；冒烟时先用 2
NUM_TRIALS_PER_TASK = 50
RUN_LIBERO_EVAL = False  # 改成 True 才会启动 500-rollout 长任务

# 选择要评测的 adapter：12-episode 实验 / 论文口径全量实验
RUN_NAME = os.environ.get('OPENVLA_RUN_NAME', 'libero_spatial_full_paper')

adapter_dir = repo / 'adapter-tmp' / RUN_NAME / (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
if not (adapter_dir / 'adapter_config.json').is_file():
    raise FileNotFoundError(f'adapter 不存在: {adapter_dir}')

environment = os.environ.copy()
environment['MUJOCO_GL'] = 'egl'                     # 离屏渲染走 EGL
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['TOKENIZERS_PARALLELISM'] = 'false'

if RUN_LIBERO_EVAL:
    result = subprocess.run(
        [
            sys.executable,
            str(repo / 'experiments/robot/libero/run_libero_eval.py'),
            '--model_family', 'openvla',
            '--pretrained_checkpoint', str(adapter_dir),
            '--base_model_path', 'openvla/openvla-7b',
            '--task_suite_name', 'libero_spatial',
            '--num_trials_per_task', str(NUM_TRIALS_PER_TASK),
            '--center_crop', 'True',
            '--load_in_4bit', 'True',
            '--seed', '7',
            '--run_id_note', RUN_NAME,
        ],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-5000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过 500-rollout LIBERO 评测；需要时将 RUN_LIBERO_EVAL 改为 True。')


## 17. LIBERO simulator 成功率结果（12-episode / 1000-step adapter）

评测设置：`libero_spatial` 全部 10 个任务 × 10 次 rollout = 100 个 episode，4-bit 量化加载 base + LoRA adapter，
`MUJOCO_GL=egl` 离屏渲染。结果日志：`experiments/logs/EVAL-libero_spatial-openvla-2026_09_21-18_53_03--12ep_1000step_10trials.txt`。

**总成功率：28.0%（28/100）**

| 任务 | 训练数据里出现过？ | 成功率 |
|---|---|---|
| 1. bowl between the plate and the ramekin | 是（1 条 episode） | **90%** |
| 2. bowl next to the ramekin | 否 | 0% |
| 3. bowl from table center | 否 | 0% |
| 4. bowl on the cookie box | 否 | 0% |
| 5. bowl in the top drawer of the wooden cabinet | 是（3 条） | 10% |
| 6. bowl on the ramekin | 是（4 条） | **60%** |
| 7. bowl next to the cookie box | 是（1 条） | 30% |
| 8. bowl on the stove | 是（3 条） | **90%** |
| 9. bowl next to the plate | 否 | 0% |
| 10. bowl on the wooden cabinet | 否 | 0% |

关键结论：**训练时出现过的 5 个任务合计 28/50 = 56%，没出现过的 5 个任务合计 0/50 = 0%。**
也就是说，28% 这个总分几乎完全由数据覆盖决定，而不是模型学不会——12 个 episode 只覆盖了 10 个任务里的 5 个
（前 12 个 episode 的指令分布：task1 ×1、task5 ×3、task6 ×4、task7 ×1、task8 ×3）。

论文口径对比需要注意：OpenVLA 论文用**完整** LIBERO-Spatial 训练集（每个任务 50 条 demo、共 500 条）微调后才报告约 85%；
本实验只用了 432 个 episode 里的前 12 个（约 2.4% 的数据量），且评测次数是 10 次/任务（论文是 50 次），因此两者不可直接比较。

后面想继续推进的话有三条路：

1. **扩大训练数据**（最有效）：把 `OPENVLA_TRAIN_EPISODES` 从 12 提到能覆盖 10 个任务的规模（例如 200），同样跑 1000 步，再评测——预期 0% 的那 5 个任务会开始得分；
2. **增加评测次数**：`--num_trials_per_task 50`，对齐论文口径（约 16 小时）；
3. **补 base 模型对照**：用未微调的 `openvla/openvla-7b` 跑同一套评测，量化微调带来的增益。


In [ ]:
from pathlib import Path

import pandas as pd
import tensorflow_datasets as tfds

# 1) 训练数据覆盖了哪些任务（train[:12] 的指令分布）
dataset_root = str(Path.home() / 'modified_libero_rlds')
train_split = tfds.load(
    'libero_spatial_no_noops',
    data_dir=dataset_root,
    split='train[:12]',
    shuffle_files=False,
)
train_coverage = {}
for example in train_split:
    step = next(iter(example['steps'].as_numpy_iterator()))
    instruction = step['language_instruction'].decode('utf-8', errors='replace')
    train_coverage[instruction] = train_coverage.get(instruction, 0) + 1

# 2) 解析最新的 LIBERO 评测日志
log_candidates = sorted((repo / 'experiments/logs').glob('EVAL-libero_spatial-*10trials*.txt'))
if not log_candidates:
    raise FileNotFoundError('没有找到 LIBERO 评测日志，请先执行第 16 节。')
log_path = log_candidates[-1]
print('评测日志:', log_path.name)

rows, current_task = [], None
for line in log_path.read_text(encoding='utf-8', errors='replace').splitlines():
    if line.startswith('Task: '):
        current_task = line[len('Task: '):].strip()
    elif line.startswith('Current task success rate:') and current_task:
        rows.append({'task': current_task, 'success_rate': float(line.split(':')[1])})
        current_task = None

table = pd.DataFrame(rows)
table['train_demos'] = table['task'].map(lambda t: train_coverage.get(t, 0))
table['success_rate_%'] = (table['success_rate'] * 100).round(1)
table = table[['task', 'train_demos', 'success_rate_%']]
print(table.to_string(index=False))

seen = table[table['train_demos'] > 0]
unseen = table[table['train_demos'] == 0]
print()
print(f"训练覆盖到的任务: {len(seen)} 个 | 平均成功率 {seen['success_rate_%'].mean():.1f}%")
print(f"训练未覆盖的任务: {len(unseen)} 个 | 平均成功率 {unseen['success_rate_%'].mean() if len(unseen) else float('nan'):.1f}%")
print(f"整体成功率: {table['success_rate_%'].mean():.1f}%")


## 18. 论文对齐实验（全量数据 + 500 次 rollout）

对照 OpenVLA 论文 Appendix E 与仓库 README 的 LIBERO 设置，本项目与论文的对应关系：

| 项目 | 论文 | 本项目 |
|---|---|---|
| 训练数据 | 官方清洗后的 LIBERO-Spatial：500 条演示中剔除 68 条失败演示 = **432 条** | `train[:432]`，同一份 `libero_spatial_no_noops` RLDS 数据 |
| 数据清洗 | 256×256 重渲染、删除 no-op 动作、第三人称图像旋转 180°、剔除失败演示 | 全部由官方 RLDS 提供（数据集名里的 `no_noops` 即第 2 项） |
| 微调方式 | LoRA r=32，作用于所有线性层 | 一致（`--lora_rank 32`、all-linear），另有 4-bit NF4 量化以适配 12GB 显存 |
| 数据增强 | 每帧随机裁剪 90% 面积 | `--image_aug true`，评测端 `center_crop` 取中心 90% |
| 评测 | 每个 suite 500 次 rollout（10 任务 × 50 次），3 个随机种子取平均 | `--num_trials_per_task 50`（1 个种子，3 个种子约需 3 倍时间） |
| 参考成绩 | LIBERO-Spatial **84.7 ± 0.9%**（论文的 3 seeds 平均） | 仅作参考，不能当作本机结果 |

论文里的微调步数没有公开，本实验按「等效 batch 16、约 1 个 epoch」选取步数：
全量数据 52,970 个 transition ÷ 16 ≈ **3,300 步/epoch**，因此采用 3,500 步（约 1.06 epoch），实测约 14.5 s/step → 约 14 小时。

运行记录：训练已经完成；正式 simulator 评测需要单独运行并通过下一节的完整性检查。

- 训练：`python reproduction/run_qlora_1000.py --train-episodes 432 --max-steps 3500 --save-steps 500 --run-name libero_spatial_full_paper`
- 每 500 步保存一次 adapter（长跑中断也能留下可用权重）
- 评测：第 20 节，`NUM_TRIALS_PER_TASK = 50`，`RUN_NAME = 'libero_spatial_full_paper'`

> **方法学边界**：本实验使用 4-bit NF4、只跑 1 个 seed，和论文的 3 seeds × 500 rollouts 并不完全相同。更准确的说法是“在本地资源约束下完成了论文设置的近似复现”，而不是声称复现了论文的 84.7 ± 0.9%。

## 19. 复现状态审计：训练完成 ≠ 评测完成

这一节只读取本地日志和权重，不会重新训练，也不会把计划中的数字当成结果。全量模型使用 `train[:432]`，因此它已经见过 `train[12:16]`；第 13 节的 12/4 held-out 动作误差只能用于 **12-episode 实验**，不能作为全量模型的 held-out 结果。

我们把项目拆成两条独立证据链：

1. **12-episode 研究实验**：`train[:12]` → `train[12:16]`，已有 MAE/MSE/L2 和 100-rollout LIBERO 结果。
2. **全量论文对齐实验**：`train[:432]` → `libero_spatial` 500 rollouts；训练权重已保存，成功率要以正式评测日志为准。


In [ ]:
from pathlib import Path
import re
import pandas as pd

full_run_name = 'libero_spatial_full_paper'
full_exp_id = (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
full_adapter = repo / 'adapter-tmp' / full_run_name / full_exp_id
full_train_log = repo / 'reproduction' / 'finetune_libero_spatial_full_paper.log'
eval_log_candidates = sorted((repo / 'experiments' / 'logs').glob(
    'EVAL-libero_spatial-openvla-*full_paper*50trials*.txt'
))

def parse_eval_log(path):
    text = path.read_text(encoding='utf-8', errors='replace') if path.exists() else ''
    task_rates = []
    for match in re.finditer(
        r'Task: (?P<task>.+?)\n.*?Current task success rate: (?P<rate>[0-9.]+)',
        text,
        flags=re.S,
    ):
        task_rates.append({'task': match.group('task').strip(), 'success_rate': float(match.group('rate'))})
    episode_counts = [int(value) for value in re.findall(r'# episodes completed so far: (\d+)', text)]
    success_counts = [int(value) for value in re.findall(r'# successes: (\d+) \(', text)]
    return {
        'path': path,
        'bytes': path.stat().st_size if path.exists() else 0,
        'task_rates': task_rates,
        'episodes': max(episode_counts, default=0),
        'successes': max(success_counts, default=0),
    }

train_text = full_train_log.read_text(encoding='utf-8', errors='replace') if full_train_log.exists() else ''
training_complete = (
    '100%|██████████| 3500/3500' in train_text
    and 'Max step 3500 reached! Stopping training...' in train_text
    and (full_adapter / 'adapter_config.json').is_file()
    and (full_adapter / 'adapter_model.safetensors').is_file()
)
eval_records = [parse_eval_log(path) for path in eval_log_candidates]
best_eval = max(eval_records, key=lambda item: (item['episodes'], item['bytes']), default=None)
formal_eval_complete = bool(
    best_eval
    and best_eval['episodes'] >= 500
    and len(best_eval['task_rates']) >= 10
)

status = pd.DataFrame([
    {'stage': '12-episode held-out action metrics', 'status': 'complete', 'evidence': str(repo / 'planB2_libero_gt.csv')},
    {'stage': '12-episode LIBERO simulator (10 × 10)', 'status': 'complete', 'evidence': '28/100 = 28.0% in 2026-09-21 log'},
    {'stage': 'full-data QLoRA training (432 episodes, 3500 steps)', 'status': 'complete' if training_complete else 'incomplete', 'evidence': str(full_train_log)},
    {'stage': 'full-data LIBERO simulator (10 × 50)', 'status': 'complete' if formal_eval_complete else 'pending', 'evidence': str(best_eval['path']) if best_eval else 'no usable 50-trial log'},
])
print(status.to_string(index=False))
print()
print('full adapter:', full_adapter)
print('adapter size (decimal MB):', round((full_adapter / 'adapter_model.safetensors').stat().st_size / 10**6, 1) if (full_adapter / 'adapter_model.safetensors').exists() else 'missing')
if best_eval:
    print('best full-paper eval log:', best_eval['path'].name)
    print('episodes parsed:', best_eval['episodes'], '| task summaries:', len(best_eval['task_rates']))
else:
    print('尚未找到可用于汇报的 full-paper simulator 评测日志。')


## 20. Full-paper LIBERO 评测：可复现运行单元（默认安全模式）

全量 adapter 已经存在，不需要重新训练。正式评测是 10 个任务 × 50 次 rollout，预计需要数小时；因此下面的 `RUN_FORMAL_EVAL` 默认是 `False`，先打印命令和检查项。确认 GPU、MuJoCo 和 LIBERO 都正常后，把它改为 `True` 再执行。

评测中断后不要读取半截日志当结果；重新启动完整 500-rollout 评测，并用第 21 节检查是否真的出现 10 个 task summary 和 500 个 episode。

In [ ]:
import os
import subprocess
import sys

RUN_FORMAL_EVAL = False  # 改成 True 才会启动长时间评测
NUM_TRIALS_PER_TASK = 50
adapter_dir = repo / 'adapter-tmp' / 'libero_spatial_full_paper' / (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
if not (adapter_dir / 'adapter_config.json').is_file():
    raise FileNotFoundError(f'full adapter 不存在: {adapter_dir}')

eval_command = [
    sys.executable, str(repo / 'experiments/robot/libero/run_libero_eval.py'),
    '--model_family', 'openvla',
    '--pretrained_checkpoint', str(adapter_dir),
    '--base_model_path', 'openvla/openvla-7b',
    '--task_suite_name', 'libero_spatial',
    '--num_trials_per_task', str(NUM_TRIALS_PER_TASK),
    '--center_crop', 'True',
    '--load_in_4bit', 'True',
    '--seed', '7',
    '--run_id_note', 'full_paper_50trials',
]
eval_env = os.environ.copy()
eval_env.update({
    'MUJOCO_GL': 'egl',
    'HF_HUB_OFFLINE': '1',
    'TRANSFORMERS_OFFLINE': '1',
    'TOKENIZERS_PARALLELISM': 'false',
})
print('adapter:', adapter_dir)
print('command:', ' '.join(eval_command))
gpu_probe = subprocess.run(['nvidia-smi', '--query-gpu=name,memory.free', '--format=csv,noheader'], text=True, capture_output=True)
print('GPU probe:', gpu_probe.stdout.strip() or gpu_probe.stderr.strip())
if RUN_FORMAL_EVAL:
    completed = subprocess.run(eval_command, cwd=repo, env=eval_env, text=True, check=False)
    print('评测退出码:', completed.returncode)
else:
    print('安全模式：没有启动长时间评测。确认资源可用后，将 RUN_FORMAL_EVAL 改为 True。')


## 21. 正式评测结果回填（只接受完整日志）

运行第 20 节后重新执行本节。代码会选择已解析 episode 数最多的 full-paper 日志，并只在 `500 episodes + 10 个 task summary` 都存在时计算总成功率。这样 notebook 不会把旧 adapter、冒烟结果或中途退出的日志混在一起。

In [ ]:
if not eval_records:
    print('没有 full-paper 评测日志；当前正式成功率仍为 pending。')
elif not formal_eval_complete:
    print(f'日志 {best_eval["path"].name} 只解析到 {best_eval["episodes"]}/500 个 episode、{len(best_eval["task_rates"])} 个 task summary。')
    print('不要把这个中间状态写成最终 success rate。')
else:
    task_table = pd.DataFrame(best_eval['task_rates'])
    task_table['success_rate_%'] = (task_table['success_rate'] * 100).round(1)
    print(task_table[['task', 'success_rate_%']].to_string(index=False))
    total_success_rate = best_eval['successes'] / best_eval['episodes']
    print()
    print(f'Full-paper success rate: {best_eval["successes"]}/{best_eval["episodes"]} = {total_success_rate:.1%}')
    print(f'日志: {best_eval["path"]}')


## 22. 结论与可诚实汇报的范围

| 结论 | 当前证据 | 能否写进简历 |
|---|---|---|
| 12-episode QLoRA 训练成功 | 1000/1000 steps，adapter 可加载 | 可以 |
| held-out 动作误差改善 | `train[12:16]` 上 MAE 0.3947 → 0.3188（−19.2%） | 可以，注明 50 transitions |
| 12-episode simulator 成功率 | 10 tasks × 10 trials：28/100 = 28.0% | 可以，作为小数据实验观察 |
| 全量 432-episode QLoRA 训练成功 | 3500/3500 steps，约 11.5 小时，adapter 约 969 MB | 可以 |
| 全量模型 500-rollout 成功率 | 只有正式日志完整后才能填写 | 目前不要填写数字 |

全量训练覆盖了 `train[:432]`，所以不能继续把 `train[12:16]` 当成全量模型的泛化集。论文的 84.7 ± 0.9% 是 3 个 seed 的平均参考值，不是本机实验结果。

## 23. 简历友好版项目总结

下面这段可以按你的实际经历改写；它只使用当前已经有证据的数字：

> **OpenVLA-7B 在 LIBERO-Spatial 上的 QLoRA 复现**
> - 基于 OpenVLA-7B 和清洗后的 LIBERO-Spatial RLDS 数据，设计 `train[:12]` / `train[12:16]` 的 episode-level held-out 协议，并实现离线缓存、单卡兼容和 adapter 验证流程。
> - 使用 4-bit NF4 量化、LoRA rank 32、有效 batch size 16 完成 1000-step 小数据实验和 432-episode / 3500-step 全量训练；最终保存约 969 MB LoRA adapter。
> - 在 50 个 held-out transitions 上将动作 MAE 从 0.3947 降至 0.3188（−19.2%）；在 10 个 LIBERO-Spatial 任务 × 10 次 rollout 上得到 28.0% 成功率，并分析出训练任务覆盖范围是主要瓶颈。
> - 编写可复现实验 notebook：自动检查数据切分、训练日志、adapter 完整性和 simulator 评测日志，区分小数据 held-out 指标与全量模型的 simulator 指标。

不要在正式 500-rollout 日志完整之前写“全量模型达到 84.7%”；那是论文参考成绩，不是这次本地运行的结果。